Core/include/Acts/Geometry/NavigationDelegates.hpp

```diff
@@ -0,0 +1,81 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Geometry/NavigationState.hpp"
+#include "Acts/Utilities/Delegate.hpp"
+
+#include <memory>
+#include <tuple>
+
+/// @note this is foreseen for the 'Geometry' module
+
+namespace Acts {
+
+class Surface;
+
+namespace Experimental {
+
+class Portal;
+class DetectorVolume;
+class Detector;
+
+/// Base class for all link implementations that need class structure
+class IDelegateImpl {
+ public:
+  virtual ~IDelegateImpl() {}
+};
+
+/// Memory managed delegate to guarantee the lifetime
+/// of eventual unterlying delegate memory and the
+/// delegate function
+///
+template <typename deletage_type>
+struct ManagedDelegate {
```

Comment:
Just to clarify, you would were thinking of:

```c++
template <typename function_type> 
struct ManagedDelegate {
  Delegate<function_type> delegate;
}
```

right?

Reply:
yeah I agree that restricts it to the `Delegate` class. when ever we see a need to generalize this we can make it more broad

Reply:
Ok, will change it.

Reply:
Mmmm, maybe it's not possible ?

```sh
/Users/salzburg/Documents/work/dev/acts-experimental/Tests/UnitTests/Core/Utilities/DelegateTests.cpp:230:19: error: 'void' is not a class type
  auto pImplRaw = dynamic_cast<const ProviderImpl*>(mpDel.implementation.get());
                  ^                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@benjaminhuth @paulgessinger

Reply:
I get the same error with `std::dynamic_pointer_cast`

```c++
auto pImplRaw = std::dynamic_pointer_cast<ProviderImpl>(mpDel.implementation);
```

Reply:
Hmm yeah so the problem is with `dynamic_cast` we need to cast from a polymorphic type to a related polymorphic type. But void is not polymorphic, so it is not possible to dynamic_cast from a `void *` to anything I would say

Reply:
Maybe we need a reinterpret cast here? but that is kind of dangerous terrain

Reply:
Ok, thanks - I need this down-casting possibility as I said e.g. for writing the delegates out and displaying them. So I will reintroduce a `IManagedDelegateImpl` base class, because a static_cast would need some sort of identification which if mismatches due to a bug does not help us.